OmsiHookInvoker/DXHook.cpp

```diff
@@ -0,0 +1,213 @@
+{
+	if (!m_device)
+		return OHERR_NOD3DDEVICE; 
+	if ((int)m_device == -2)
+		return OHERR_D3DDEVICEQUERYFAILED;
+
+	// Shared handle is only supported with D3D9Ex sadly:
+	// https://www.gamedev.net/forums/topic/638495-shared-resources-eg-textures-between-devicesthreads/5030524/
+	
+	return m_device->CreateTexture(Width, Height, 1, D3DUSAGE_DYNAMIC, Format, D3DPOOL_DEFAULT, ppTexture, NULL);
+}
+
+HRESULT DXHook::UpdateSubresource(IDirect3DTexture9* Texture, UINT8* TextureData, UINT Width, UINT Height, BOOL UseRect, LONG32 Left, LONG32 Top, LONG32 Right, LONG32 Bottom)
+{
+	// Check the arguments, users can never be trusted and this is annoying to debug...
+	if (Texture == nullptr || TextureData == nullptr)
+		return OHERR_TEXTURENULL;
+
+    D3DSURFACE_DESC desc;
+    CHECK_FAILURE_RETURN(Texture->GetLevelDesc(0, &desc));
+	/*IDirect3D9* d3d;
+	CHECK_FAILURE_RETURN(m_device->GetDirect3D(&d3d));
+	D3DDISPLAYMODE displayMode;
+	CHECK_FAILURE_RETURN(d3d->GetAdapterDisplayMode(D3DADAPTER_DEFAULT, &displayMode));
+	CHECK_FAILURE_RETURN(d3d->CheckDeviceFormat(D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL, displayMode.Format, desc.Usage, D3DRTYPE_TEXTURE, desc.Format));*/
+
+	if (desc.Width < Width || desc.Height < Height)
+		return OHERR_UPDATESUBRES_DSTTEXTURETOOSMALL;
+
+	if (UseRect)
+		if (Right - Left != Width || Bottom - Top != Height)
+			return OHERR_UPDATESUBRES_INVALIDRECT;
+
+	const RECT rect {
+		Left,
+		Top,
+		Right,
+		Bottom
+	};
+	D3DLOCKED_RECT lockedRect;
+	CHECK_FAILURE_RETURN(Texture->LockRect(0, &lockedRect, UseRect ? &rect : NULL, D3DLOCK_DISCARD));
+
+    int bpp = BitsPerPixel(desc.Format) / 8;
+	for (UINT y = 0; y < Height; y++)
+	{
+		void* dst = ((UINT8*)lockedRect.pBits) + (y * lockedRect.Pitch);
+		memcpy(dst, (TextureData + y * Width * bpp), Width * bpp);
+	}
+
+	CHECK_FAILURE_RETURN(Texture->UnlockRect(0));
+
+	return S_OK;
+}
+
+HRESULT DXHook::ReleaseTexture(IDirect3DTexture9* Texture)
+{
+    if (Texture == nullptr)
+        return OHERR_TEXTURENULL;
+    return Texture->Release();
+}
+
+HRESULT DXHook::GetTextureDesc(IDirect3DTexture9* Texture, UINT* pWidth, UINT* pHeight, UINT* pFormat)
+{
+    if (Texture == nullptr)
+        return OHERR_TEXTURENULL;
+    D3DSURFACE_DESC desc;
+    CHECK_FAILURE_RETURN(Texture->GetLevelDesc(0, &desc));
+    *pWidth = desc.Width;
+    *pHeight = desc.Height;
+    *pFormat = desc.Format;
+    return S_OK;
+}
+
+HRESULT DXHook::IsTexture(IUnknown* Texture)
+{
+    if (Texture == nullptr)
+        return OHERR_TEXTURENULL;
+    IDirect3DTexture9* itex;
+    return Texture->QueryInterface(&itex);
+}
+
+//-------------------------------------------------------------------------------------
+// Returns bits-per-pixel for a given D3D9 format, or 0 on failure
+//-------------------------------------------------------------------------------------
+constexpr size_t DXHook::BitsPerPixel(D3DFORMAT fmt) noexcept
```

Comment:
is this not a copy of the switch from the other class?

Reply:
Yes it mirrors the C# implementation, that is intentional.